### PR TITLE
osutil/disks: introduce RegisterDeviceMapperBackResolver + use for crypt-luks2

### PR DESCRIPTION
**Pull request by anonymouse64**
This is broken out from #11028, which itself is actually broken out from #10851 originally.

Add a new callback functionality for registering handlers which "un-map" a
device mapper node back to the original source for purposes of creating and/or
locating a disk device.

The CRYPT-LUKS2 based implementation currently supported is now implemented as
a handler this way, and this paves the way for additional handlers to be
implemented as well, specifically in other packages so that the osutil/disks
package can remain as generic as possible and easily imported from anywhere
without import cycles.

### CODE REVIEW

File: osutil/disks/disks_linux.go [hunk 1]
```diff
@@ -448,52 +435,29 @@ func diskFromPartUDevProps(props map[string]string, opts *Options) (*disk, error
...
-
-		// now finally, we need to use this uuid, which is the device uuid of
-		// the actual physical encrypted partition to get the path, which will
-		// be something like /dev/vda4, etc.
-		byUUIDPath := filepath.Join("/dev/disk/by-uuid", canonicalUUID)
-		props, err = udevPropertiesForName(byUUIDPath)
-		if err != nil {
-			return nil, fmt.Errorf("cannot get udev properties for encrypted partition %s: %v", byUUIDPath, err)
+		if !matchedHandler {
+			return nil, fmt.Errorf("no handler supports decrypted device mapper with UUID %q and name %q", dmUUID, dmName)
```
Comment (pedronis) [inline osutil/disks/disks_linux.go:460]: is this an internal error in practice?

internal error: no back resolver ...

if it's not an internall error, then it needs to be formulated as a "cannot back resolve " or something

**Review by pedronis — APPROVED** [review]
lgtm with some minor remarks

**Review by bboozzoo — APPROVED** [review]
LGTM with comments from @pedronis

**Review by mvo5 — APPROVED** [review]
Very nice, thank you

File: osutil/disks/disks_linux.go (hunk 1 above)
Comment (anonymouse64) [reply osutil/disks/disks_linux.go:460]: I think for now we can leave it as an internal error, at least if someone ends up using a device with a mapper that we don't support, the error message they get will be clear to us we need to add a new format back resolver whereas a non-internal error may suggest something else. But at the end of the day it doesn't really matter much either way I think

File: osutil/disks/luks.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	// systemd-cryptsetup when unlocking
+	// the actual value that is returned also has "-some-name" appended to this
+	// pattern, but we delete that from the string before matching with this
+	// regexp to prevent issues like a mapper volume name that has CRYPT-LUKS2-
+	// in the name and thus we might accidentally match it
+	// see also the comments in DiskFromMountPoint about this value
+	luksUUIDPatternRe = regexp.MustCompile(`^CRYPT-LUKS2-([0-9a-f]{32})$`)
+)
+
+func cryptLuks2DeviceMapperHandler(dmUUID, dmName []byte) (dev string, ok bool) {
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply osutil/disks/luks.go]: done

File: osutil/disks/disks.go [hunk 3]
```diff
@@ -187,3 +187,24 @@ func (e PartitionNotFoundError) Error() string {
 var (
 	_ = error(PartitionNotFoundError{})
 )
+
+var deviceMapperBackResolvers = map[string]func(dmUUID, dmName []byte) (dev string, ok bool){}
+
+// RegisterDeviceMapperHandler takes a callback function which is used when the
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply osutil/disks/disks.go]: done

**Pull request merged**